beacon-chain/sync/initial-sync/blocks_fetcher.go

```diff
@@ -397,76 +392,127 @@ func lowestSlotNeedsBlob(retentionStart primitives.Slot, bwb []blocks2.BlockWith
 	if bwb[len(bwb)-1].Block.Block().Slot() < retentionStart {
 		return nil
 	}
-	for _, b := range bwb {
+	fc := make([]commitmentCount, 0, len(bwb))
+	for i := range bwb {
+		b := bwb[i]
 		slot := b.Block.Block().Slot()
+		if b.Block.Version() < version.Deneb {
+			continue
+		}
 		if slot < retentionStart {
 			continue
 		}
 		commits, err := b.Block.Block().Body().BlobKzgCommitments()
 		if err != nil || len(commits) == 0 {
 			continue
 		}
-		return &slot
+		fc = append(fc, commitmentCount{slot: slot, root: b.Block.Root(), count: len(commits)})
 	}
-	return nil
+	return fc
 }
 
-func sortBlobs(blobs []blocks.ROBlob) []blocks.ROBlob {
-	sort.Slice(blobs, func(i, j int) bool {
-		if blobs[i].Slot() == blobs[j].Slot() {
-			return blobs[i].Index < blobs[j].Index
+func slotRangeForCommitmentCounts(fc []commitmentCount, bs filesystem.BlobStorageSummarizer) *blobRange {
+	if len(fc) == 0 {
+		return nil
+	}
+	// If we don't have a blob summarizer, can't check local blobs, request blobs over complete range.
+	if bs == nil {
+		return &blobRange{low: fc[0].slot, high: fc[len(fc)-1].slot}
+	}
+	for i := range fc {
+		hci := fc[i]
+		if bs.Summary(hci.root).AllAvailable(hci.count) {
+			continue
 		}
-		return blobs[i].Slot() < blobs[j].Slot()
-	})
+		// We found a missing blob
+		needed := &blobRange{low: hci.slot, high: hci.slot}
+		for z := len(fc) - 1; z > i; z-- {
+			hcz := fc[z]
+			if bs.Summary(hcz.root).AllAvailable(hcz.count) {
+				continue
+			}
+			needed.high = hcz.slot
+			return needed
+		}
```

Comment:
Does this assume that fc is a sorted list? If so, is that a safe assumption?

Reply:
It's a requirement per specs: https://github.com/prysmaticlabs/prysm/blob/develop/beacon-chain/sync/rpc_send_request.go#L93-L103